instancetype :  Add `InstancetypeResourcesRef` to ` InstancetypeStatusRef `

lyarwood opened the pull request:
### What this PR does

This adds a new `InstancetypeStatusResources` struct containing CPU and
Memory fields to `InstancetypeStatusRef`, exposed via the `vm.status.instancetypeRef.resources` JSON
field. This allows users and tooling to see the guest visible resources provided by a
referenced instancetype directly in the VirtualMachine status without
needing to run the VM, fetch the instancetype object, or call expand-spec.

Resources reflect the **currently active** instancetype. When a running VM has
a `RestartRequired` condition set (pending instancetype change that requires
restart), the resources field preserves the current values until the restart
occurs.

```
$ virtctl create vm --instancetype u1.medium --name test | kubectl apply -f -
virtualmachine.kubevirt.io/test created
$ kubectl get vms/test -o json | jq .status.instancetypeRef
{
  "controllerRevisionRef": {
    "name": "test-u1.medium-v1beta1-cf448630-a54d-4323-97ab-c407f0d99079-1"
  },
  "name": "u1.medium",
  "resources": {
    "cpu": {
      "cores": 1,
      "sockets": 1,
      "threads": 1
    },
    "memory": "4Gi"
  }
}

$ kubectl patch vms/test --type='merge' -p '{"spec":{"instancetype":{"name": "u1.large"}}}'
virtualmachine.kubevirt.io/test patched
kubectl get vms/test -o json | jq .status.instancetypeRef
{
  "controllerRevisionRef": {
    "name": "test-u1.large-v1beta1-9f64d9ee-9bb0-4097-8b19-8e6bfc96bf47-1"
  },
  "name": "u1.large",
  "resources": {
    "cpu": {
      "cores": 1,
      "sockets": 2,
      "threads": 1
    },
    "memory": "8Gi"
  }
}
```

### References

- VEP 280: https://github.com/kubevirt/enhancements/issues/280
- VEP PR: https://github.com/kubevirt/enhancements/pull/281
- Fixes # https://issues.redhat.com/browse/CNV-62832

### Why we need it and why it was done in this way

The resources field is populated during the existing `Store()` flow in
`storeInstancetypeRevision`. When the instancetype reference changes,
resources are recalculated from the new instancetype — unless the VM is
running with a `RestartRequired` condition, in which case the existing
resources are preserved to reflect the currently active instancetype.

After restart (or successful hotplug migration), `RestartRequired` is cleared
and the next reconcile recalculates resources from the now-active instancetype.

### Special notes for your reviewer

- No feature gate — this is a read-only, optional status field with no behavioral changes
- Preferences influence the CPU topology breakdown (cores/sockets/threads) and
  are applied during resources calculation
- The `RestartRequired` preservation logic is covered by dedicated unit tests

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present ([VEP 278](https://github.com/kubevirt/enhancements/pull/281))
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Guest visible vCPU and memory resources provided by a referenced instance type are now visible under `vm.status.instancetypeRef.resources.{cpu,memory}`
```

lyarwood committed:
api(instancetype): Add InstancetypeStatusResources to InstancetypeStatusRef

This adds a new InstancetypeStatusResources struct containing CPU and
Memory fields to InstancetypeStatusRef, exposed via the "resources" JSON
field. This allows users and tooling to see the resources provided by a
referenced instancetype directly in the VirtualMachine status without
needing to run the VM, fetch the instancetype object, or call expand-spec.

Signed-off-by: Lee Yarwood <lyarwood@redhat.com>

lyarwood committed:
instancetype: Wire up InstancetypeStatusResources

Signed-off-by: Lee Yarwood <lyarwood@redhat.com>

lyarwood committed:
instancetype: Add tests to ensure CR not created when revisionName set

Signed-off-by: Lee Yarwood <lyarwood@redhat.com>

lyarwood committed:
chore(api): Run gofumpt on virtv1 types.go

Signed-off-by: Lee Yarwood <lyarwood@redhat.com>

lyarwood commented on pkg/instancetype/revision/store.go:
```diff
@@ -140,6 +135,13 @@ func (h *revisionHandler) storeInstancetypeRevision(vm *virtv1.VirtualMachine) (
 	}
 
+	// Populate InstancetypeStatusResources if not already set
+	if statusRef.InstancetypeStatusResources == nil {
```
Yeah I can add some comments here, this is required when no changes have been made to the matcher or a revisionName is provided by the matcher. In both cases we don't fetch objects and create a CR thus the need to handle that here.

iholder101 commented:
Hey @lyarwood! Thanks for this PR!

I think the work done here is somewhat overlapping with VEP 156: https://github.com/kubevirt/enhancements/pull/162.

The general idea there is to expose the original memory and CPU requests (still debating the exact API). I think it will serve your goal but would be more generalized, as every VM (not just ones that use instance type) would expose this information.

Would you mind having a look on the VEP and share your thoughts?

lyarwood commented:
> Hey @lyarwood! Thanks for this PR!
> 
> I think the work done here is somewhat overlapping with VEP 156: [kubevirt/enhancements#162](https://github.com/kubevirt/enhancements/pull/162).
> 
> The general idea there is to expose the original memory and CPU requests (still debating the exact API). I think it will serve your goal but would be more generalized, as every VM (not just ones that use instance type) would expose this information.
> 
> Would you mind having a look on the VEP and share your thoughts?

IMHO that's unrelated. It's specifically against VMIs where guest visible vCPU and memory resources are already rendered into the spec when using a VM with an instance type. 

This PR specifically targets the VM where we currently have to use the expand-spec API or start the VM to see these fully rendered values.

lyarwood commented:
/reopen

lyarwood commented:
Rebased onto main and addressed the following review feedback:

- **Fixed swapped store arguments in `populateInstancetypeStatusResources`**: The `instancetypeStore`/`clusterInstancetypeStore` and `preferenceStore`/`clusterPreferenceStore` arguments to `NewSpecFinder` were swapped, causing lookups to use the wrong stores.
- **Renamed `InstancetypeStatusResources` field to `Resources`** on `InstancetypeStatusRef`: The Go field name now matches the JSON tag `resources`, eliminating the `names_match` API rule violation. Updated across all generated files, tests, and implementation code.
- **Fixed pre-existing test bug**: The `revision.New()` call in `store_test.go` was passing `clusterInstancetypeInformerStore` instead of `clusterPreferenceInformerStore` as the 4th argument.
- **Moved CNV-9098 test assertions to `tests/instancetype/revisions.go`** to match the upstream test restructuring.

lyarwood commented on pkg/instancetype/revision/store.go:
```diff
@@ -148,6 +155,30 @@ func (h *revisionHandler) storeInstancetypeRevision(vm *virtv1.VirtualMachine) (
 	return vm.Status.InstancetypeRef, nil
 }
 
+func (h *revisionHandler) populateInstancetypeStatusResources(
+	vm *virtv1.VirtualMachine,
+	statusRef *virtv1.InstancetypeStatusRef,
+) error {
+	vmApplier := apply.NewVMApplier(
+		find.NewSpecFinder(h.instancetypeStore, h.clusterInstancetypeStore, h.revisionStore, h.virtClient),
+		preferenceFind.NewSpecFinder(h.preferenceStore, h.clusterPreferenceStore, h.revisionStore, h.virtClient),
+	)
+	vmCopy := vm.DeepCopy()
+	if err := vmApplier.ApplyToVM(vmCopy); err != nil {
```
Added nil checks for both `domain.CPU` and `domain.Memory.Guest` after `ApplyToVM`, returning a clear error if either is not defined.

lyarwood commented on pkg/instancetype/revision/store_test.go:
```diff
@@ -185,6 +191,9 @@ var _ = Describe("Instancetype and Preferences revision handler", func() {
 				Expect(vm.Status.InstancetypeRef.Name).To(Equal(vm.Spec.Instancetype.Name))
 				Expect(vm.Status.InstancetypeRef.Kind).To(Equal(vm.Spec.Instancetype.Kind))
 				Expect(vm.Status.InstancetypeRef.ControllerRevisionRef.Name).To(Equal(clusterInstancetypeControllerRevision.Name))
+				Expect(vm.Status.InstancetypeRef.Resources).ToNot(BeNil())
+				Expect(vm.Status.InstancetypeRef.Resources.CPU.Sockets).To(Equal(clusterInstancetype.Spec.CPU.Guest))
+				Expect(vm.Status.InstancetypeRef.Resources.Memory).To(Equal(clusterInstancetype.Spec.Memory.Guest))
```
Added `CPU.Cores` and `CPU.Threads` assertions to all unit test assertion blocks.

lyarwood committed:
instancetype: Preserve resources during RestartRequired and add missing tests

Resources in vm.status.instancetypeRef should reflect the currently active
instancetype, not a pending one. When a running VM has RestartRequired set,
skip resources recalculation to preserve the current values until the
restart occurs.

Also adds unit tests for stopped VM resources population, instancetype
switching on stopped VMs, RestartRequired resources preservation, and
post-restart resources update.

Assisted-By: Claude <noreply@anthropic.com>
Signed-off-by: Lee Yarwood <lyarwood@redhat.com>

dhiller commented:
/test pull-kubevirt-e2e-k8s-1.36-sig-compute-migrations
